### PR TITLE
CVSL-395: Refactor caseloadService

### PR DESCRIPTION
**Pull request by smcveigh941**
- Refactor caseloadService to allow caseloads configurable by release date and licence status. 
- OMU Caseload shows urgent NOT_STARTED licences
- All caseloads now return the same information in the same format - common interface
- Each caseload can be configured more easily by release date and statuses

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -76,6 +76,7 @@ jobs:
           root: .
           paths:
             - node_modules
+            - assets/stylesheets
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply .circleci/config.yml:79]: Well spotted!

File: integration_tests/mockApis/licence.ts [hunk 2]
```diff
@@ -6,7 +6,7 @@ const licencePlaceholder = {
   typeCode: 'AP_PSS',
   version: '1.1',
   statusCode: 'IN_PROGRESS',
-  nomsId: 'A1234AA',
+  nomsId: 'G9786GC',
   bookingNo: '123456',
```
Comment (timharrison-moj) [inline integration_tests/mockApis/licence.ts:10]: Will this affect some of the other integration tests?

File: server/routes/approvingLicences/handlers/approvalCases.ts [hunk 3]
```diff
@@ -13,20 +15,20 @@ export default class ApprovalCaseRoutes {
     const caseloadViewModel = cases
       .map(c => {
         return {
-          licenceId: c.licenceId,
-          name: `${c.forename} ${c.surname}`.trim(),
-          prisonNumber: c.nomisId,
-          probationPractitioner: `${c.comFirstName} ${c.comLastName}`.trim(),
-          releaseDate: moment(c.conditionalReleaseDate, 'DD/MM/YYYY').format('DD MMM YYYY'),
+          licenceId: _.head(c.licences).id,
+          name: convertToTitleCase(`${c.nomisRecord.firstName} ${c.nomisRecord.lastName}`.trim()),
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/routes/approvingLicences/handlers/approvalCases.ts:19]: Not sure I understand why licences is a list at this point, but assume this is part of the caseload changes, so reading on..

[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/routes/approvingLicences/handlers/approvalCases.ts:19]: Update: I see. Top-level components of a caseload item are `licence`, `nomisRecord`, `deliusRecord` and `probationPractioner` - all of which are lists returned from a map-like function, even if there is only one licence in the list.

File: server/routes/varyingLicences/handlers/caseload.ts [hunk 4]
```diff
@@ -11,29 +14,32 @@ export default class CaseloadRoutes {
...
       ? await this.caseloadService.getTeamVaryCaseload(user)
       : await this.caseloadService.getStaffVaryCaseload(user)
 
-    const caseloadViewModel = licences
-      .map(licence => {
+    const caseloadViewModel = cases
+      .map(c => {
+        const licence =
+          c.licences.length > 1 ? c.licences.find(l => l.status !== LicenceStatus.ACTIVE) : _.head(c.licences)
+
```
Comment (timharrison-moj) [inline server/routes/varyingLicences/handlers/caseload.ts:25]: Under what circumstances would an individual have a licences list > 1 long? Does the caseload service return all licences for a person together as a list?

Comment (timharrison-moj) [reply server/routes/varyingLicences/handlers/caseload.ts:25]: Is the find !== ACTIVE sufficient to rule out other status codes that might not want to be shown in the vary caseload? e.g. INACTIVE (or APPROVED, and in future, DISCARDED).

File: server/routes/viewingLicences/index.ts [hunk 5]
```diff
@@ -33,10 +33,10 @@ export default function Index({
   const comDetailsHandler = new ComDetailsRoutes(communityService)
 
   get('/cases', viewCasesHandler.GET)
+  get('/probation-practitioner/staffCode/:staffCode', comDetailsHandler.GET)
   get('/id/:licenceId/show', viewLicenceHandler.GET)
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/routes/viewingLicences/index.ts:37]: Actually, it looks like the probation practitioner display is still duplicated in each route handler file across create, vary, approve,

File: server/routes/approvingLicences/handlers/approvalCases.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (smcveigh941) [reply server/routes/approvingLicences/handlers/approvalCases.ts:19]: An offender may have multiple licences if they have one which is ACTIVE, and one which is e.g. VARIATION_IN_PROGRESS. The caseloadService returns a list and its up to the consumer to handle it. There is some logic in the vary caseload handler to choose the variation if one exists and ignore the active one

File: server/routes/varyingLicences/handlers/caseload.ts (hunk 4 above)
Comment (smcveigh941) [reply server/routes/varyingLicences/handlers/caseload.ts:25]: Someone may have both an ACTIVE and VARIATION_IN_PROGRESS licence at the same time. The vary caseload builder in caseloadService would not return APPROVED or DISCARDED licences.

File: server/routes/viewingLicences/handlers/viewCases.test.ts [hunk 6]
```diff
@@ -33,28 +31,56 @@ describe('Route handlers - View and print case list', () => {
...
+        probationPractitioner: {
+          name: 'Thor',
+        },
+      },
+      {
+        licences: [
+          {
+            type: LicenceType.AP,
+            status: LicenceStatus.NOT_IN_PILOT,
+          },
```
[the remark this answers is not in the document]
Comment (smcveigh941) [reply server/routes/viewingLicences/handlers/viewCases.test.ts:73]: Just testing that the NOT_IN_PILOT status (which is if the licence is not started but they are not in the rollout area) is shown on the caseload view

File: server/routes/viewingLicences/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (smcveigh941) [reply server/routes/viewingLicences/index.ts:37]: I switched to use staffCode in order to be able to build that url without the need for an extra API call to get the staffIdentifier... staffCode is readily available, getting the staffIdentifier requires another call. Is it a problem to expose it? I thought they didnt want to expose IDs and move towards using staffCode instead anyway?

File: jobs/promptLicenceCreation.ts [hunk 7]
```diff
@@ -26,22 +26,24 @@ const pollPrisonersDueForLicence = async (): Promise<Prisoner[]> => {
...
-            prisoner =>
-              prisoner.licenceStatus === LicenceStatus.NOT_STARTED ||
-              prisoner.licenceStatus === LicenceStatus.IN_PROGRESS
+        prisoners.filter(offender =>
+          [LicenceStatus.NOT_STARTED, LicenceStatus.IN_PROGRESS].some(status =>
+            offender.licences.find(l => l.status === status)
           )
-          .map(prisoner => prisoner.nomisRecord)
+        )
       )
```
[the remark this answers is not in the document]
Comment (smcveigh941) [reply jobs/promptLicenceCreation.ts:41]: Yes trying to reuse as much as possible in this job

File: server/routes/creatingLicences/types/additionalConditionsYesOrNo.ts [hunk 8]
```diff
@@ -2,7 +2,7 @@ import { Expose } from 'class-transformer'
 import { IsIn, IsNotEmpty } from 'class-validator'
 import YesOrNo from '../../../enumeration/yesOrNo'
 
-const message = 'Select yes or no for whether they need additional licence conditions'
+const message = 'Select yes or not applicable'
```
[the remark this answers is not in the document]
Comment (smcveigh941) [reply server/routes/creatingLicences/types/additionalConditionsYesOrNo.ts:6]: Oops, mistake!

File: server/routes/viewingLicences/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/routes/viewingLicences/index.ts:37]: I think its ok - not very meaningful to anyone.

**Review by timharrison-moj — APPROVED** [review]
LGTM. Hold off merge until demo's done? After 2pm today I think.

**Pull request merged**